File: src/main/java/com/joecollins/bindings/NestedBindableList.java
package com.joecollins.bindings;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.WeakHashMap;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

public class NestedBindableList<T extends Bindable<E>, E extends Enum<E>> extends BindableList<T> {
  private Map<E, List<BiConsumer<Integer, T>>> itemBindings = new HashMap<>();
  private Map<BiConsumer<Integer, T>, Map<Integer, Consumer<T>>> storedBindings =
      new WeakHashMap<>();

  public void addNestedItemBinding(BiConsumer<Integer, T> binding, E... properties) {
    addItemBinding(binding);
    for (E property : properties)
      itemBindings.computeIfAbsent(property, x -> new LinkedList<>()).add(binding);
    for (int index = 0; index < size(); index++) {
      T item = get(index);
      if (item != null) {
        item.addBinding(getOrCreateBinding(binding, index), properties);
        for (E property : properties) item.onPropertyRefreshed(property);
      }
    }
  }

  public void removeNestedItemBinding(BiConsumer<Integer, T> binding, E... properties) {
    removeItemBinding(binding);
    for (E property : properties)
      itemBindings.computeIfAbsent(property, x -> new LinkedList<>()).remove(binding);
    for (int index = 0; index < size(); index++) {
      T item = get(index);
      if (item != null) {
        item.removeBinding(getOrCreateBinding(binding, index), properties);
      }
    }
  }

  @Override
  protected void addAllBindings(int index) {
    super.addAllBindings(index);
    T item = get(index);
    for (Map.Entry<E, List<BiConsumer<Integer, T>>> entry : itemBindings.entrySet()) {
      E property = entry.getKey();
      List<BiConsumer<Integer, T>> consumers = entry.getValue();
      for (BiConsumer<Integer, T> consumer : consumers) {
        if (item != null) {
          item.addBinding(getOrCreateBinding(consumer, index), property);
        }
      }
    }
  }

  @Override
  protected void removeAllBindings(int index) {
    super.removeAllBindings(index);
    T item = get(index);
    for (Map.Entry<E, List<BiConsumer<Integer, T>>> entry : itemBindings.entrySet()) {
      E property = entry.getKey();
      List<BiConsumer<Integer, T>> consumers = entry.getValue();
      for (BiConsumer<Integer, T> consumer : consumers) {
        if (item != null) {
          item.removeBinding(getOrCreateBinding(consumer, index), property);
        }
      }
    }
  }

  private Consumer<T> getOrCreateBinding(BiConsumer<Integer, T> consumer, int index) {
    return storedBindings
        .computeIfAbsent(consumer, x -> new HashMap<>())
        .computeIfAbsent(index, i -> (val -> consumer.accept(i, val)));
  }
}

File: src/main/java/com/joecollins/bindings/NestedBindableList.kt
package com.joecollins.bindings

import java.util.LinkedList
import java.util.WeakHashMap
import java.util.function.Consumer
import kotlin.collections.ArrayList
import kotlin.collections.HashMap

class NestedBindableList<T : Bindable<T, E>, E : Enum<E>>(items: List<T> = ArrayList()) : BindableList<T>(items) {

    private val itemBindings: MutableMap<E, MutableList<(Int, T) -> Unit>> = HashMap()
    private val storedBindings: MutableMap<(Int, T) -> Unit, MutableMap<Int, Consumer<T>>> = WeakHashMap()

    fun addNestedItemBinding(binding: (Int, T) -> Unit, vararg properties: E) {
        addItemBinding(binding)
        properties.forEach { property ->
            itemBindings.computeIfAbsent(property, { LinkedList() }).add(binding)
        }
        (0 until size).forEach { index ->
            val item = get(index)
            item.addBinding(getOrCreateBinding(binding, index), *properties)
            for (property in properties) item.onPropertyRefreshed(property)
        }
    }

    fun removeNestedItemBinding(binding: (Int, T) -> Unit, vararg properties: E) {
        removeItemBinding(binding)
        properties.forEach { property ->
            itemBindings.computeIfAbsent(property, { LinkedList() }).remove(binding)
        }
        (0 until size).forEach { index ->
            val item = get(index)
            item.removeBinding(getOrCreateBinding(binding, index), *properties)
        }
    }

    override fun addAllBindings(index: Int) {
        super.addAllBindings(index)
        val item = get(index)
        itemBindings.forEach { (property, consumers) ->
            for (consumer in consumers) {
                item.addBinding(getOrCreateBinding(consumer, index), property)
            }
        }
    }

    override fun removeAllBindings(index: Int) {
        super.removeAllBindings(index)
        val item = get(index)
        itemBindings.forEach { (property, consumers) ->
            for (consumer in consumers) {
                item.removeBinding(getOrCreateBinding(consumer, index), property)
            }
        }
    }

    private fun getOrCreateBinding(consumer: (Int, T) -> Unit, index: Int): Consumer<T>? {
        return storedBindings
            .computeIfAbsent(consumer, { HashMap() })
            .computeIfAbsent(index) { idx ->
                Consumer { value: T ->
                    consumer(idx, value)
                }
            }
    }
}
